File: src/main/java/com/onepointsixtwo/dagger_viewmodel_processor/InjectorSubclass.java
package com.onepointsixtwo.dagger_viewmodel_processor;


import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;

/**
 * Creates a specific subclass of injector for injecting a specific activity or fragment
 */
public class InjectorSubclass extends BaseClass {

    private String name;
    private TypeSpec.Builder injectorBuilder;
    private TypeName enclosingClassType;
    private VariableElement viewModelFieldInEnclosingClass;
    private BaseInjectorClass baseInjectorClass;
    private InjectViewModelProcessor.Type type;
    private boolean useActivityScope;

    public InjectorSubclass(TypeElement enclosingClassType,
            BaseInjectorClass baseInjectorClass,
            VariableElement viewModelFieldInEnclosingClass,
            InjectViewModelProcessor.Type enclosingClassTypeEnum,
            boolean useActivityScope) {
        this.baseInjectorClass = baseInjectorClass;
        this.viewModelFieldInEnclosingClass = viewModelFieldInEnclosingClass;
        this.type = enclosingClassTypeEnum;
        this.useActivityScope = useActivityScope;

        this.enclosingClassType = ClassName.get(enclosingClassType);
        name = enclosingClassType.getSimpleName() + "ViewModelInjector";

        createBasicClass();
        addInjectMethod();
    }

    private void createBasicClass() {
        injectorBuilder = TypeSpec.classBuilder(name)
                .superclass(baseInjectorClass.classType())
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL);
    }

    private void addInjectMethod() {
        TypeName viewModelType = ClassName.get(viewModelFieldInEnclosingClass.asType());
        TypeName viewModelProviders = ClassName.bestGuess("androidx.lifecycle.ViewModelProviders");

        String methodName = "inject";
        String injecteeVariableName = "injectee";
        String factoryVariableName = "factory";

        MethodSpec.Builder injectMethodBuilder = MethodSpec.methodBuilder(methodName)
                .addModifiers(Modifier.PUBLIC)
                .addParameter(javaObjectType, injecteeVariableName)
                .addParameter(factoryType, factoryVariableName)
                .addAnnotation(Override.class);

        if (type == InjectViewModelProcessor.Type.FRAGMENT_TYPE && useActivityScope) {
            injectMethodBuilder.addStatement("(($T)$L).$L = $T.of((($T)$L).getActivity(), $L).get($T.class)",
                    enclosingClassType,
                    injecteeVariableName,
                    viewModelFieldInEnclosingClass.getSimpleName(),
                    viewModelProviders,
                    enclosingClassType,
                    injecteeVariableName,
                    factoryVariableName,
                    viewModelType);
        } else {
            injectMethodBuilder.addStatement("(($T)$L).$L = $T.of(($T)$L, $L).get($T.class)",
                    enclosingClassType,
                    injecteeVariableName,
                    viewModelFieldInEnclosingClass.getSimpleName(),
                    viewModelProviders,
                    enclosingClassType,
                    injecteeVariableName,
                    factoryVariableName,
                    viewModelType);
        }

        MethodSpec injectMethod = injectMethodBuilder.build();

        injectorBuilder.addMethod(injectMethod);
    }

    @Override
    protected TypeSpec.Builder getTypeBuilder() {
        return injectorBuilder;
    }

    @Override
    TypeName classType() {
        return ClassName.bestGuess(name);
    }
}

File: src/main/java/com/onepointsixtwo/dagger_viewmodel_processor/InjectorSubclass.kt
package com.onepointsixtwo.dagger_viewmodel_processor


import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.TypeSpec

import javax.lang.model.element.TypeElement
import javax.lang.model.element.VariableElement

/**
 * Creates a specific subclass of injector for injecting a specific activity or fragment
 */
class InjectorSubclass(
    enclosingClassType: TypeElement,
    private val baseInjectorClass: BaseInjectorClass,
    private val viewModelFieldInEnclosingClass: VariableElement,
    private val type: InjectViewModelProcessor.Type,
    private val useActivityScope: Boolean
) : BaseClass() {

    private val name: String
    protected override var typeBuilder: TypeSpec.Builder? = null
        private set
    private val enclosingClassType: TypeName

    init {

        this.enclosingClassType = ClassName.bestGuess(enclosingClassType.qualifiedName.toString())
        name = enclosingClassType.simpleName.toString() + "ViewModelInjector"

        createBasicClass()
        addInjectMethod()
    }

    private fun createBasicClass() {
        typeBuilder = TypeSpec.classBuilder(name)
            .superclass(baseInjectorClass.classType())
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
    }

    private fun addInjectMethod() {
        val viewModelType = ClassName.bestGuess(viewModelFieldInEnclosingClass.asType().toString())
        val viewModelProviders = ClassName.bestGuess("androidx.lifecycle.ViewModelProviders")

        val methodName = "inject"
        val injecteeVariableName = "injectee"
        val factoryVariableName = "factory"

        val injectMethodBuilder = FunSpec.builder(methodName)
            .addModifiers(KModifier.PUBLIC)
            .addParameter(injecteeVariableName, javaObjectType)
            .addParameter(factoryVariableName, factoryType)
            .addAnnotation(Override::class.java)

        if (type == InjectViewModelProcessor.Type.FRAGMENT_TYPE && useActivityScope) {
            injectMethodBuilder.addStatement(
                "((\$T)\$L).\$L = \$T.of(((\$T)\$L).getActivity(), \$L).get(\$T.class)",
                enclosingClassType,
                injecteeVariableName,
                viewModelFieldInEnclosingClass.simpleName,
                viewModelProviders,
                enclosingClassType,
                injecteeVariableName,
                factoryVariableName,
                viewModelType
            )
        } else {
            injectMethodBuilder.addStatement(
                "((\$T)\$L).\$L = \$T.of((\$T)\$L, \$L).get(\$T.class)",
                enclosingClassType,
                injecteeVariableName,
                viewModelFieldInEnclosingClass.simpleName,
                viewModelProviders,
                enclosingClassType,
                injecteeVariableName,
                factoryVariableName,
                viewModelType
            )
        }

        val injectMethod = injectMethodBuilder.build()

        typeBuilder!!.addFunction(injectMethod)
    }

    internal override fun classType(): TypeName {
        return ClassName.bestGuess(name)
    }

    override fun fileName(): String {
        return name
    }
}
